Signal/src/view controllers/ExperienceUpgradeViewController.swift

```diff
@@ -0,0 +1,292 @@
+        dismissButton.autoPinEdge(toSuperviewEdge: .bottom, withInset: ScaleFromIPhone5(32))
+    }
+
+    private func buildSlideView(header: String, body: String, image: UIImage?) -> UIView {
+        Logger.debug("\(TAG) in \(#function)")
+
+        let containerView = UIView()
+        let headerView = UIView()
+        containerView.addSubview(headerView)
+        headerView.backgroundColor = UIColor.ows_materialBlue()
+
+        headerView.autoPinWidthToSuperview()
+        headerView.autoPinEdge(toSuperviewEdge: .top, withInset: -16)
+        headerView.autoSetDimension(.height, toSize: ScaleFromIPhone5(100))
+
+        // Title label
+        let titleLabel = UILabel()
+        headerView.addSubview(titleLabel)
+        titleLabel.text = header
+        titleLabel.textAlignment = .center
+        titleLabel.font = UIFont.ows_regularFont(withSize: ScaleFromIPhone5To7Plus(26, 32))
+        titleLabel.textColor = UIColor.white
+        titleLabel.minimumScaleFactor = 0.5
+        titleLabel.adjustsFontSizeToFitWidth = true;
+
+        // Title label layout
+        titleLabel.autoPinWidthToSuperview(withMargin: ScaleFromIPhone5To7Plus(16, 24))
+        titleLabel.autoPinEdge(toSuperviewEdge: .bottom, withInset: ScaleFromIPhone5(16))
+
+        let slideView = UIView()
+        containerView.addSubview(slideView)
+
+        let containerPadding = ScaleFromIPhone5To7Plus(12, 24)
+        slideView.autoPinEdgesToSuperviewEdges(with: UIEdgeInsets(top: containerPadding, left: containerPadding, bottom: containerPadding, right: containerPadding))
+
+        // Body label
+        let bodyLabel = UILabel()
+        slideView.addSubview(bodyLabel)
+        bodyLabel.text = body
+        bodyLabel.font = UIFont.ows_lightFont(withSize: ScaleFromIPhone5To7Plus(18, 22))
+        bodyLabel.textColor = UIColor.black
+        bodyLabel.numberOfLines = 0
+        bodyLabel.textAlignment = .center
+
+        // Body label layout
+        bodyLabel.autoPinWidthToSuperview()
+        bodyLabel.sizeToFit()
+
+        // Image
+        let imageView = UIImageView(image: image)
+        slideView.addSubview(imageView)
+        imageView.contentMode = .scaleAspectFit
+
+        // Image layout
+        imageView.autoPinWidthToSuperview()
+        imageView.autoSetDimension(.height, toSize: ScaleFromIPhone5To7Plus(200, 280))
+        imageView.autoPinEdge(.top, to: .bottom, of: headerView, withOffset: ScaleFromIPhone5To7Plus(24, 32))
+        imageView.autoPinEdge(.bottom, to: .top, of: bodyLabel, withOffset: -ScaleFromIPhone5To7Plus(24, 32))
+        
+        return containerView
+    }
+
+    private func addDismissGesture() {
+        let swipeGesture = UISwipeGestureRecognizer(target: self, action: #selector(handleDismissGesture))
+        swipeGesture.direction = .down
+        view.addGestureRecognizer(swipeGesture)
+    }
+
+    // MARK: Carousel
+
+    private var carouselView: UIScrollView!
+    private var currentSlideIndex = 0
```

Comment:
Huh, so you didn't want to gather these at the top of the class.

Did you consider using `UIPageViewController`?  It solves a very similar problem pretty well.